### PR TITLE
Allow any pytree

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
# PR: Arbitrary Pytree Support for X Input

## Summary

Extend the test suite and documentation to formally cover arbitrary JAX pytrees as the `X`
argument to all GLM models, and fix two source-level issues discovered during testing.

## Motivation

`FeaturePytree` was the only officially documented and tested path for grouped feature input.
JAX's own pytree mechanism is more general (dicts, `equinox.Module`, nested containers, etc.)
and avoids bespoke wrapper types. This PR validates that the implementation already supports
arbitrary pytrees, removes `FeaturePytree` from every X-parameter description, and adds
regression coverage so future refactors cannot silently break pytree inputs.

### Source fixes

**`base_validator.py` — `validate_inputs`**: replaced `X.shape[0]` with a loop over 
`jax.tree_util.tree_leaves(X)` to obtain n_samples from an arbitrary pytree. Also added
a consistency check that all leaves share the same number of samples.

**`validation.py`**: updated structure-mismatch error messages to report
`jax.tree_util.tree_structure(X)` vs `jax.tree_util.tree_structure(coef)` instead of
assuming X is an array; 

**`glm.py` — `initialize_params`**: added a `feature_mask_consistency` call so that a
stale array `_feature_mask` set by a previous `fit(array_X, y)` raises a clear `TypeError`
instead of silently producing wrong shapes when the model is re-initialized with a pytree X. Added a test for this new behavior, see below. This was not necessary for the pytree support but that I caught this as I was writing new tests.

### Tests

| Layer | File | What is tested |
|---|---|---|
| 1 | `test_proximal_operator.py` | Each of the 5 prox operators accepts a pytree of params; array-vs-tree equivalence |
| 2 | `test_regularizer.py` | `_SIMPLE_PYTREE_CASES` uses plain dict / eqx module / dict-list instead of FeaturePytree |
| 3 | `test_base_regressor_subclasses.py` | `TestModelVsPytree` — all public API methods involving X × models × pytree-type |
| 4 | `test_glm.py` | Coef tree structure matches X after `initialize_params` and `fit`; stale mask regression |

`test_initialize_params_stale_feature_mask_raises` guards the source fix above: before this
PR the mask/pytree mismatch would surface at the first loss evaluation, producing a cryptic
shape error far from the site of the mistake.

### Documentation

- Removed every X-parameter reference to `FeaturePytree` in method docstrings across
  `GLM`, `PopulationGLM`, `ClassifierGLM`, and `ClassifierPopulationGLM`; replaced with
  "pytree of arrays of the same shape".
- Updated `PopulationGLM` class docstring: replaced the "Mask Coefficients with a
  FeaturePytree" example with a plain-dict equivalent.
- Added a "Use a Pytree of Arrays as Input" example to the class docstrings of `GLM`,
  `ClassifierGLM`, and `ClassifierPopulationGLM`.

## Open Questions for the Reviewers

- **FeaturePytree backward compatibility**. I would want to drop the FeaturePytree implementation entirely at this point. In a subsequent PR I would like to add a deprecation warning. For how many versions should we be backward compatible?
- What should I do with the showcasing the FeaturePytree? (Probably nothing in this one but in the next PR deprecating the feature pytree)
  - **plot_03_population_glm.md**: This shows that the `feature_mask` is automatically generated based on the tree structure, one group per leaf. I think this should stay and be replaced with a regular dictionary.
  - **plot_07_glm_pytree.md**: This one literally shows the FeaturePytree on its own. Should we show the pytree compatibility with different tree types instead or should we just delete it?
  - **finegrained_regularization.md**: This shows how to regularize with structured regularization strength, as long as the strength structure matches that of the parameters. I could just show how to do it with dict and list of coefficients.
  - **variable_selection_group_lasso.md**: This had a note on how to structure the regularization by grouping variables (using FeaturePytree for now). What should I do here? If I keep `plot_07_glm_pytree` using other tree types, I can keep the link to it.

### CODE REVIEW

File: src/nemos/glm/classifier_glm.py [hunk 1]
```diff
@@ -1026,6 +1037,16 @@ class labels from the provided ``y``. If you set ``coef_`` and ``intercept_`` ma
     ... )
     >>> model.regularizer
     Ridge()
+
+    **Use a Pytree of Arrays as Input**
```
Comment (wulfdewolf) [inline src/nemos/glm/classifier_glm.py]: uncapitalise 'Arrays', to be consistent

File: src/nemos/glm/classifier_glm.py [hunk 2]
```diff
@@ -762,6 +762,17 @@ class labels from the provided ``y``. If you set ``coef_`` and ``intercept_`` ma
...
```
Comment (wulfdewolf) [inline src/nemos/glm/classifier_glm.py]: uncapitalise 'Arrays', to be consistent

File: src/nemos/glm/glm.py [hunk 3]
```diff
@@ -229,6 +229,17 @@ class GLM(BaseRegressor[GLMUserParams, GLMParams]):
     >>> model = nmo.glm.GLM(solver_name="LBFGS").fit(X, y)
     >>> model.solver_name
     'LBFGS[...]'
+
+    **Use a Pytree of Arrays as Input**
```
Comment (wulfdewolf) [inline src/nemos/glm/glm.py]: uncapitalise 'Arrays', to be consistent

File: src/nemos/glm/glm.py [hunk 4]
```diff
@@ -384,7 +395,8 @@ def predict(self, X: DESIGN_INPUT_TYPE) -> jnp.ndarray:
         Parameters
         ----------
         X :
-            Predictors, array of shape ``(n_time_bins, n_features)`` or pytree of same.
+            Predictors, array of shape ``(n_time_bins, n_features)`` or pytree of the same
```
Comment (wulfdewolf) [inline src/nemos/glm/glm.py]: or a pytree of arrays of the same shape.

to be consistent

File: src/nemos/glm/glm.py [hunk 5]
```diff
@@ -502,7 +514,8 @@ def score(
         Parameters
         ----------
         X :
-            The exogenous variables. Shape ``(n_time_bins, n_features)``.
+            Predictors, array of shape ``(n_time_bins, n_features)`` or pytree of the same
+            shape.
```
Comment (wulfdewolf) [inline src/nemos/glm/glm.py]: or a pytree of arrays of the same shape.

to be consistent

File: tests/conftest.py [hunk 6]
```diff
@@ -69,14 +73,17 @@ def initialize_feature_mask_for_population_glm(X, n_neurons: int, coef=None):
         of shape (n_neurons,) for each key.
         If X is an array, returns an array of shape (n_features, n_neurons).
     """
+    extra_shape = (n_classes,) if n_classes else ()
     if coef is not None:
         return jax.tree_util.tree_map(lambda c: jnp.ones(c.shape), coef)
     if isinstance(X, FeaturePytree):
```
Comment (wulfdewolf) [inline tests/conftest.py:79]: this is kept for backward compability then?

**Review by wulfdewolf — COMMENTED** [review]
LGTM, some minor text comments in docstrings.

Regarding the documentation:
- `plot_03_population_glm.md`: I agree, keep but replace with `dict` pytrees
- `plot_07_glm_pytree.md`: I like the idea of having a dedicated little notebook to first of all explain the basic concept of a pytree and then also how NeMoS interacts with them
- `finegrained_regularization.md`: I agree, replace with `dict` pytrees
- `variable_selection_group_lasso.md`: I agree, keep, reference in the pytree notebook

File: tests/conftest.py (hunk 6 above)
Comment (BalzaniEdoardo) [reply tests/conftest.py:79]: yes, for now. It is very cool to see that dropping the FeaturePytree gets rid of the conditional logic and makes the code simpler and more general
